tensorflow/core/kernels/dml_strided_slice_op.cc

```diff
@@ -501,4 +504,251 @@ TF_CALL_int8(REGISTER_KERNEL);
 TF_CALL_int64(REGISTER_KERNEL);
 #undef REGISTER_KERNEL
 
+class StridedSliceAssignInitHelper : public InitializationHelper {
+ public:
+  struct Attributes {
+    explicit Attributes(OpKernelConstruction* ctx) {
+      OP_REQUIRES_OK(ctx, ctx->GetAttr("begin_mask", &begin_mask));
+      OP_REQUIRES_OK(ctx, ctx->GetAttr("end_mask", &end_mask));
+      OP_REQUIRES_OK(ctx, ctx->GetAttr("ellipsis_mask", &ellipsis_mask));
+      OP_REQUIRES_OK(ctx, ctx->GetAttr("new_axis_mask", &new_axis_mask));
+      OP_REQUIRES_OK(ctx, ctx->GetAttr("shrink_axis_mask", &shrink_axis_mask));
+    }
+
+    int32 begin_mask, end_mask;
+    int32 ellipsis_mask, new_axis_mask, shrink_axis_mask;
+  };
+
+  bool IsNoOpKernel(
+      OpKernelContext* ctx,
+      absl::Span<const TensorShape> output_shapes) const override {
+    // StridedSliceAssign is only a no-op if the first input or its output is
+    // empty
+    return ctx->input(0).NumElements() == 0 ||
+           (!output_shapes.empty() && output_shapes[0].num_elements() == 0);
+  }
+
+  StridedSliceAssignInitHelper(OpKernelContext* ctx,
+                               std::shared_ptr<const Attributes> attr) {
+    DCHECK(ctx->input_is_ref(0) || ctx->input(0).dtype() == DT_RESOURCE);
+
+    if (ctx->input(0).dtype() == DT_RESOURCE) {
+      OP_REQUIRES_OK(
+          ctx, LookupResource(ctx, HandleFromInput(ctx, 0), &input_resource_));
+      input_resource_->mu()->lock_shared();
```

Comment:
Should this be unlocked if an exception is thrown in one of the OP_REQUIRES_OK calls below? I suspect TF won't be able to recover regardless, but the destructor won't unlock this if there's a validation error. Might be best to use a cleanup like you did in compute and reset it if we get to the end of the constructor successfully.